### ISSUE TITLE
C: Mount Point not Correctly Determined in WSL 2

### ISSUE THREAD

**Issue by BinToss** [tillig/ps-bash-completions#10]
`/bin/bash: c/Users/NoahR/OneDrive/Documents/PowerShell/Modules/PSBashCompletions/1.2.3/bash_completion_bridge.sh: No such file or directory`
This path may be valid for Git-Bash, but not for WSL's Bash.

WSL mount points have a `mnt` parent directory e.g. `/mnt/c/Users/NoahR`

WSL's Bash.exe is typically found at `$env:LOCALAPPDATA\Microsoft\WindowsApps\bash.exe`

----

```powershell
C:\Users\NoahR> where.exe bash
C:\Users\NoahR\AppData\Local\Microsoft\WindowsApps\bash.exe
C:\Users\NoahR> Test-Path $env:LOCALAPPDATA\Microsoft\WindowsApps\bash.exe
True
```

**Comment by tillig** [comment]
If you would, could you possibly expand the explanation of the challenge here? Terse is fine; this is maybe a little too terse. I'm not sure if you're having a problem with the path to `bash_completion_bridge.sh` or that `bash` itself isn't being located.

You mentioned something about `/mnt` being part of it, [there is code to try to detect where the mount point is](https://github.com/tillig/ps-bash-completions/blob/24fb9fbfcc878c2ce0ef92852e17d4bfa809766a/PSBashCompletions/PSBashCompletions.psm1#L126) but perhaps that's not working in some of these later WSL instances.

**Comment by BinToss** [comment]
I was under the impression that the Powershell module didn't support `/mnt/` and I'm very puzzled now.
I wonder why it decided to look under `/c/` instead of `/mnt/c/`.

**Comment by tillig** [comment]
OK, so the problem you're seeing isn't that it's not finding bash, it's that it isn't finding the bridge script. (Please confirm this - I really do need that clarification I asked for. Sorry to be pedantic about that.)

The code to find that script, as you see in that link, is:

`$mountPath = (&"$bash" -c "mount | grep -e '^C:\\\\\\?[[:space:]]' | cut -f3 -d\ ") -replace "/c", "/"`

The way I'd debug it is to run each of the command subsets in turn and see the output.

In bash, what does `mount` yield for you?

What about `mount | grep -e '^C:\\\\\\?[[:space:]]'` ?

Now the whole thing - `mount | grep -e '^C:\\\\\\?[[:space:]]' | cut -f3 -d\ `

(There's an escaped trailing space in that last one - don't forget it!)

We need to break down which part is yielding something unexpected.

**Comment by BinToss** [comment]
`mount | grep -e '^C:\\\\\\?[[:space:]]'` outputs nothing so $mountPath is empty.
`mount` outputs the following
```
none on /mnt/wsl type tmpfs (rw,relatime)
drivers on /usr/lib/wsl/drivers type 9p (ro,nosuid,nodev,noatime,dirsync,aname=drivers;fmask=222;dmask=222,mmap,access=client,msize=65536,trans=fd,rfd=7,wfd=7)
none on /usr/lib/wsl/lib type overlay (rw,relatime,lowerdir=/gpu_lib_packaged:/gpu_lib_inbox,upperdir=/gpu_lib/rw/upper,workdir=/gpu_lib/rw/work)
/dev/sdc on / type ext4 (rw,relatime,discard,errors=remount-ro,data=ordered)
none on /mnt/wslg type tmpfs (rw,relatime)
/dev/sdc on /mnt/wslg/distro type ext4 (ro,relatime,discard,errors=remount-ro,data=ordered)
rootfs on /init type rootfs (rw,size=8104976k,nr_inodes=2026244)
none on /dev type devtmpfs (rw,nosuid,relatime,size=8105004k,nr_inodes=2026251,mode=755)
sysfs on /sys type sysfs (rw,nosuid,nodev,noexec,noatime)
proc on /proc type proc (rw,nosuid,nodev,noexec,noatime)
devpts on /dev/pts type devpts (rw,nosuid,noexec,noatime,gid=5,mode=620,ptmxmode=000)
none on /run type tmpfs (rw,nosuid,nodev,mode=755)
none on /run/lock type tmpfs (rw,nosuid,nodev,noexec,noatime)
none on /run/shm type tmpfs (rw,nosuid,nodev,noatime)
none on /run/user type tmpfs (rw,nosuid,nodev,noexec,noatime,mode=755)
binfmt_misc on /proc/sys/fs/binfmt_misc type binfmt_misc (rw,relatime)
tmpfs on /sys/fs/cgroup type tmpfs (rw,nosuid,nodev,noexec,relatime,mode=755)
cgroup2 on /sys/fs/cgroup/unified type cgroup2 (rw,nosuid,nodev,noexec,relatime,nsdelegate)
cgroup on /sys/fs/cgroup/cpuset type cgroup (rw,nosuid,nodev,noexec,relatime,cpuset)
cgroup on /sys/fs/cgroup/cpu type cgroup (rw,nosuid,nodev,noexec,relatime,cpu)
cgroup on /sys/fs/cgroup/cpuacct type cgroup (rw,nosuid,nodev,noexec,relatime,cpuacct)
cgroup on /sys/fs/cgroup/blkio type cgroup (rw,nosuid,nodev,noexec,relatime,blkio)
cgroup on /sys/fs/cgroup/memory type cgroup (rw,nosuid,nodev,noexec,relatime,memory)
cgroup on /sys/fs/cgroup/devices type cgroup (rw,nosuid,nodev,noexec,relatime,devices)
cgroup on /sys/fs/cgroup/freezer type cgroup (rw,nosuid,nodev,noexec,relatime,freezer)
cgroup on /sys/fs/cgroup/net_cls type cgroup (rw,nosuid,nodev,noexec,relatime,net_cls)
cgroup on /sys/fs/cgroup/perf_event type cgroup (rw,nosuid,nodev,noexec,relatime,perf_event)
cgroup on /sys/fs/cgroup/net_prio type cgroup (rw,nosuid,nodev,noexec,relatime,net_prio)
cgroup on /sys/fs/cgroup/hugetlb type cgroup (rw,nosuid,nodev,noexec,relatime,hugetlb)
cgroup on /sys/fs/cgroup/pids type cgroup (rw,nosuid,nodev,noexec,relatime,pids)
cgroup on /sys/fs/cgroup/rdma type cgroup (rw,nosuid,nodev,noexec,relatime,rdma)
cgroup on /sys/fs/cgroup/misc type cgroup (rw,nosuid,nodev,noexec,relatime,misc)
none on /mnt/wslg/versions.txt type overlay (rw,relatime,lowerdir=/systemvhd,upperdir=/system/rw/upper,workdir=/system/rw/work)
none on /mnt/wslg/doc type overlay (rw,relatime,lowerdir=/systemvhd,upperdir=/system/rw/upper,workdir=/system/rw/work)
none on /tmp/.X11-unix type tmpfs (rw,relatime)
drvfs on /mnt/c type 9p (rw,noatime,dirsync,aname=drvfs;path=C:\;uid=1000;gid=1000;symlinkroot=/mnt/,mmap,access=client,msize=262144,trans=virtio)
drvfs on /mnt/d type 9p (rw,noatime,dirsync,aname=drvfs;path=D:\;uid=1000;gid=1000;symlinkroot=/mnt/,mmap,access=client,msize=262144,trans=virtio)
drvfs on /mnt/g type 9p (rw,noatime,dirsync,aname=drvfs;path=G:\;uid=1000;gid=1000;symlinkroot=/mnt/,mmap,access=client,msize=262144,trans=virtio)
drvfs on /mnt/h type 9p (rw,noatime,dirsync,aname=drvfs;path=H:\;uid=1000;gid=1000;symlinkroot=/mnt/,mmap,access=client,msize=262144,trans=virtio)
```

**Comment by tillig** [comment]
I'm going to take a pause here and wait for the confirmation I've asked for in the first line of my first comment...

> If you would, could you possibly expand the explanation of the challenge here? Terse is fine; this is maybe a little too terse. I'm not sure if you're having a problem with the path to `bash_completion_bridge.sh` or that `bash`  itself isn't being located.

..._and_ in the first line of my second comment...

> OK, so the problem you're seeing isn't that it's not finding bash, it's that it isn't finding the bridge script. (Please confirm this - I really do need that clarification I asked for. Sorry to be pedantic about that.)

**Comment by BinToss** [comment]
Bash is located. The bridge script is *not* located.

**Comment by tillig** [comment]
Excellent, thanks for the confirmation. You'd be surprised how many times I've gone down a troubleshooting track just to find out I was troubleshooting the wrong thing due to miscommunication.

**Comment by tillig** [comment]
This doesn't seem to be limited to Ubuntu 20.04. It's been a couple of years since I put this module out and it's possible that things changed in WSL2 in how `mount` returns. (I'm primarily on Mac nowadays so I haven't noticed this.)

The `grep` in there is looking for a `C:\` mount, but, at least on my other laptop running Windows 10 and WSL 2, neither one reports anything that'd match that `grep` with all the backslashes - something I think was a WSL1 artifact. I don't have a WSL1 box around anymore to test with.

I think the key line now will be this one:

`drvfs on /mnt/c type 9p (rw,noatime,dirsync,aname=drvfs;path=C:\;uid=1000;gid=1000;symlinkroot=/mnt/,mmap,access=client,msize=262144,trans=virtio)`

That is, `grep` for `path=C:\` and trim from there.

Try this in bash:

`mount | grep -e 'path=C:\\' | cut -f3 -d\ `

(with the trailing escaped space)

When I run that I end up with `/mnt/c` as a result.

However, in Git bash, `mount` shows

`C: on /c type ntfs (binary,noacl,posix=0,user,nomount,auto)`

And I can't _just_ grep for `C:` because there are other mount points like

`C:/Program Files/Git on / type ntfs (binary,noacl,auto)`

that get reported.

Maybe the key is to search for `/c` and `/mnt/c` and guess from there. Any line that has both `C:` and one of those two things would be the indicator. I'll have to think about how best to calculate that. Likely it'd be execute bash to get the `mount` contents then do a little PowerShell to filter it down.

**Comment by tillig** [comment]
Releasing the fix as v1.2.4 shortly.
